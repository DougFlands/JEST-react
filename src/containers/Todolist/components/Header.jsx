import React, { Component } from "react";
// import { connect } from "react-redux";

class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: ""
    };
  }

  handleInputChange(e) {
    this.setState({
      value: e.target.value
    });
  }

  handleInputKeyUp(e) {
    if (e.keyCode === 13 && this.state.value) {
      this.props.addUndoItem({
        type: "div",
        value: this.state.value
      });
      this.setState({
        value: ""
      });
    }
  }

  render() {
    // const { value, handleInputChange } = this.props;
    const { value } = this.state;
    return (
      <div
        data-test="Header"
        style={{
          display: "flex",
          height: "24px",
          justifyContent: "space-around"
        }}
      >
        <p>todolist</p>
        <input
          data-test="input"
          value={value}
          // onChange={e => handleInputChange(e.target.value)}
          onChange={this.handleInputChange.bind(this)}
          onKeyUp={this.handleInputKeyUp.bind(this)}
        />
      </div>
    );
  }
}

// const mapState = () => {
//   return {
//     value: state.todo.inputValue
//   }
// }

// const mapDispatch = dispatch => ({
//   handleInputChange(value) {
//     dispatch(actions.changeInputValue(value))
//   }
// })

// export default connect(mapState, mapDispatch)(Header);
export default Header;
