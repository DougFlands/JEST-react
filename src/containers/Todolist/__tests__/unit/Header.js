import React from 'react';
import { shallow } from 'enzyme';
import Header from '../../components/Header'


describe('Header 组件', () => {
  it('样式渲染正常', () => {
    const wrapper = shallow(<Header />)
    expect(wrapper).toMatchSnapshot()
  })

  it('组件包含 input', () => {
    const wrapper = shallow(<Header />)
    const inputElem = wrapper.find("[data-test='input']")
    expect(inputElem.length).toBe(1)
  })

  it('input 初始化内容为空', () => {
    const wrapper = shallow(<Header />)
    const inputElem = wrapper.find("[data-test='input']")
    // 对Input做的测试，所以是prop
    expect(inputElem.prop('value')).toEqual('')
  })

  it('input 用户输入时会跟随变化', () => {
    const wrapper = shallow(<Header />)
    const inputElem = wrapper.find("[data-test='input']")
    inputElem.simulate('change', {
      target: {
        value: 'test'
      }
    })
    expect(wrapper.state('value')).toEqual('test')
    const newInputElem = wrapper.find("[data-test='input']")
    expect(newInputElem.prop('value')).toEqual('test')
  })

  it('input 用户输入回车时，无内容则无操作', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Header addUndoItem={fn} />)
    const inputElem = wrapper.find("[data-test='input']")
    wrapper.setState({ value: '' })
    inputElem.simulate('keyUp', {
      keyCode: 13
    })
    expect(fn).not.toHaveBeenCalled()
    expect(wrapper.state('value')).toEqual('')
  })

  it('input 用户输入回车时，有内容则传出，并清除', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Header addUndoItem={fn} />)
    const inputElem = wrapper.find("[data-test='input']")
    wrapper.setState({ value: 'test' })
    inputElem.simulate('keyUp', {
      keyCode: 13
    })
    expect(fn).toHaveBeenCalled()
    expect(wrapper.state('value')).toEqual('')
    // 最后一次调用时的参数
    expect(fn).toHaveBeenLastCalledWith({ "type": "div", "value": "test" })
  })


})


