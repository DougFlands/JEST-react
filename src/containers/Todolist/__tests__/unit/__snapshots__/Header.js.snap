// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header 组件 样式渲染正常 1`] = `
<div
  data-test="Header"
  style={
    Object {
      "display": "flex",
      "height": "24px",
      "justifyContent": "space-around",
    }
  }
>
  <p>
    todolist
  </p>
  <input
    data-test="input"
    onChange={[Function]}
    onKeyUp={[Function]}
    value=""
  />
</div>
`;
